src/Core/Framework/Adapter/Cache/InvalidatorStorage/RedisInvalidatorStorage.php

```diff
@@ -27,14 +29,65 @@ public function store(array $tags): void
 
     public function loadAndDelete(): array
     {
-        /** @var array{0: list<string>, 1: mixed} $values */
-        $values = $this
-            ->redis
-            ->multi()
-            ->sMembers(self::KEY)
-            ->del(self::KEY)
-            ->exec();
-
-        return $values[0];
+        $tags = $this->loadAndDeleteMultiTransaction();
+
+        if ($tags !== null) {
+            return $tags;
+        }
+
+        return $this->loadAndDeleteSequentialFallback();
+    }
+
+    /**
+     * @return list<string>|null
+     */
+    private function loadAndDeleteMultiTransaction(): ?array
+    {
+        try {
+            /** @var array{0: list<string>, 1: mixed}|false $values */
+            $values = $this
+                ->redis
+                ->multi()
+                ->sMembers(self::KEY)
+                ->del(self::KEY)
+                ->exec();
+
+            if ($values !== false) {
+                return $values[0];
+            }
+        } catch (\Throwable $e) {
+            $this->logger->warning('Redis transaction failed, falling back to sequential execution. Error: ' . $e->getMessage());
+
+            return null;
+        }
+
+        $this->logger->warning('Redis transaction failed (exec returned false), falling back to sequential execution.');
+
+        return null;
+    }
+
+    /**
+     * @return list<string>
+     */
+    private function loadAndDeleteSequentialFallback(): array
+    {
+        // This breaks atomicity but ensures the queue is drained
+        try {
+            $tags = [];
+
+            $chunk = $this->redis->sPop(self::KEY, 10000);
+            while (\is_array($chunk) && !empty($chunk)) {
+                foreach ($chunk as $tag) {
+                    $tags[] = (string) $tag;
+                }
+                $chunk = $this->redis->sPop(self::KEY, 10000);
+            }
+        } catch (\Throwable $e) {
+            $this->logger->error('Sequential fallback: Could not load and delete tags from Redis. Error: ' . $e->getMessage());
+
+            throw $e;
+        }
+
+        return $tags;
```

Comment:
```suggestion
                $tags[] = $chunk;
                $chunk = $this->redis->sPop(self::KEY, 10000);
            }
        } catch (\Throwable $e) {
            $this->logger->error('Sequential fallback: Could not load and delete tags from Redis. Error: ' . $e->getMessage());

            throw $e;
        }

        return array_values(array_merge([], ...$tags));
```

this is probably a bit faster, than iterating in every loop 10k pieces

Reply:
not tested! pseudo code! but you get the idea :wink: